### PR TITLE
[SE-0096] Implement type(of:)

### PR DESCRIPTION
**Pull request by CodaFi**
#### What's in this pull request?

In the interest of getting this patch in quickly (and because I already have the work done) this patch is going to supersede #3871.
#### Resolved bug number: ([SR-2218](https://bugs.swift.org/browse/SR-2218))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
:shipit:

**Comment by rxwei** [comment]
Cool 👍  Are we keeping compatibility with `x.dynamicType`? And we may want to highlight `type` and `of` as a keyword, when they appear together.

**Comment by CodaFi** [comment]
This change is source breaking.  It will be compatible until the next patch where I yank support for `dynamicType` out of the compiler 😎.

File: lib/Parse/ParseExpr.cpp [hunk 1]
```diff
@@ -2995,3 +3016,76 @@ Parser::parseVersionConstraintSpec() {
...
+
+  // Parse the leading '('.
+  SourceLoc lParenLoc = consumeToken(tok::l_paren);
+
+  // Parse `of` label.
+  auto ofRange = Tok.getRange();
+  if (Tok.canBeArgumentLabel() && peekToken().is(tok::colon)) {
+    bool hasOf = Tok.getText() == "of";
+    if (!hasOf) {
+      // User mis-spelled the 'of' label.
```
Comment (jtbandes) [inline lib/Parse/ParseExpr.cpp]: Does this mean it's impossible for user code to define a function named `type` (with any argument label)?

Comment (CodaFi) [inline lib/Parse/ParseExpr.cpp]: It does.  It is a limitation of the SE that deserves to be discussed.  One solution would be to make this a "magic hash-prefix" function like #selector.

Comment (CodaFi) [inline lib/Parse/ParseExpr.cpp]: Originally we wanted this function defined in stdlib and not in the parser.  But that can't happen for bureaucratic reasons related to protocols and other nominal types having differing expressions allowed in these dynamicType contexts.

Comment (CodaFi) [inline lib/Parse/ParseExpr.cpp]: Nevertheless, I think I'm going to disable this diagnostic when I do the next patch to properly remove the keyword make the parser fall back to find "overloads" here.

**Pull request merged**